### DISCUSSION TITLE
Question about hessians for custom operations from C++

### DISCUSSION THREAD

**Discussion by samuelpmish** [jax-ml/jax#17550] [Q&A] [answered]
Hi,

I'm interested in using JAX for finite element simulation. Some of the building blocks of finite elements (e.g. interpolation, integration) are expensive linear operators. I have an existing C++/CUDA library with optimized implementations of these operations that I was hoping to use from within JAX (rather than rewrite everything from scratch), but many of the guides I see related to C++ extensions and custom derivatives only describe how to register a VJP and JVP for these custom operations. 

Are custom operations from C++ limited to first derivatives, or is there a way to let JAX know that certain operations are linear (so that any higher order derivatives are identically zero)? I love JAX's ability to easily compute hessians, and would like to still be able to take two derivatives of code using the custom C++ interpolation / integration implementations.

Thanks

**Comment by jakevdp** [comment]
If you want your custom kernels to be compatible with JAX autodiff, you should register `custom_vjp` and/or `custom_jvp` rules for them (see [Custom derivative rules for JAX-transformable Python functions](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html)). As long as the jvp/vjp rules are themselves compatible with autodiff, then you will be order to perform second-order autodiff on these functions.

**Reply by samuelpmish** [reply]
Thank you for the response, but it didn't quite address what I'm trying to find out, so I'll try to clarify:

The custom kernel implementation is essentially a black box (i.e. a C++ library), so it's not traceable by JAX directly. However, I know that inside my C++ black box is a simple linear operator, and I've also got C++ functions in that library for its jvp and vjp. 

It's not clear to me how to communicate this to JAX, though. https://github.com/google/jax/issues/3996 makes it sounds like I'm allowed to specify exactly one of `custom_vjp` or `custom_jvp` for my custom kernel, which means I can't support both forward and reverse mode differentiation (which would be a bummer). Hessians are also a problem because the `custom_jvp` / `custom_vjp` implementations would also be calling through to a black box, so they also aren't compatible with autodiff.

--------

Put another way: how does JAX implement something like an inverse FFT? It's an enormous nonsymmetric linear operator with an optimized implementation living in a C++ library. JAX must have some mechanism for encoding which custom kernels to call for both the `vjp` and `jvp` of that operator (since it's not traceable, one can't figure out how to implement `vjp` from the black box that implements the `jvp`), and that the hessian of an `ifft` is identically zero, but how do I do the same?

**Reply by jakevdp** [reply]
Thanks for the details!

It sounds like you need two different mechanisms working together: first, you can use [`pure_callback`](https://jax.readthedocs.io/en/latest/_autosummary/jax.pure_callback.html) to wrap your C++ kernel so it's callable by JAX, and so similarly wrap your `jvp` and `vjp` kernels.

Second, you can use `custom_jvp`/`custom_vjp` on the first kernel to tell JAX to use the jvp/vjp kernels when computing its autodiff. There is a relevant example at [Pure Callback with custom_jvp](https://jax.readthedocs.io/en/latest/notebooks/external_callbacks.html#example-pure-callback-with-custom-jvp).

If you want to compute second-order autodiff, then you'll also have to define a custom jvp or vjp for your jvp/vjp kernels mentioned above. That could be via another C++ callback, or via a JAX expression if possible.

Does that make sense?

**Reply by samuelpmish** [reply]
Thank you for the suggestions, I had never heard of `pure_callback` before!

I wrote up a small C++ [example project](https://github.com/samuelpmish/jax_cpp_interop/blob/main/ext.cpp) with python bindings to try out `pure_callback`, but it doesn't seem to be able to invoke the python bindings: 

```py
# func(x) := {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}} * {x[0], x[1], x[2]}
# vjp(x) := {x[0], x[1], x[2]} * {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}}
from my_cpp_library import func, vjp                                    
                                                                                    
x = np.array([1.0, 1.0, 1.0], f64)                                                  
                                                                                    
# verify that python bindings work on jax.numpy arrays        
print(func(x)) # prints [ 6. 15. 24.] ✅                                                                       
print(vjp(x))  # prints [12. 15. 18.] ✅                                                                              
                                                                                                                                                     
def f(x):                                                                               
  result_shape_dtype = jax.ShapeDtypeStruct(shape=(3,), dtype=f64)                  
  return jax.pure_callback(func, result_shape_dtype, x)

# but calling `func` from pure_callback is problematic
print(f(x)) 

# > jaxlib.xla_extension.XlaRuntimeError: INTERNAL: Generated function failed: 
# > CpuCallback error: TypeError: func(): incompatible function arguments. The following argument types are supported:
# >   1. func(x: jaxlib.xla_extension.DeviceArray[dtype=float64, shape=(3)]) -> jaxlib.xla_extension.DeviceArray[dtype=float64, shape=(3)]
# >
# > Invoked with types: ndarray
```

Although it's ~~possible~~ probable I'm just using `pure_callback` incorrectly here.

But `pure_callback` only supporting CPU execution seems like a deal breaker as the C++/CUDA library is mostly GPU kernels (which have zero-copy interop directly with jax arrays, but aren't differentiable out of the box).

---------

I just tried searching through JAX for how it interacts with cuFFT for inspiration, but I don't really see anything in the repo. So maybe the mechanism for using libraries like cuFFT in JAX is more subtle than I expected.

**Reply by jakevdp** [reply]
It looks like `pure_callback` expects a JAX array or tracer as input. You're passing a numpy array. You could either explicitly convert `x` to jax array with `jnp.asarray()`, or wrap your function in `jit`.

**Reply by samuelpmish** [reply]
I'm getting the same error message after changing it to

```py
def f(x):
  x = jnp.asarray(x)
  result_shape_dtype = jax.ShapeDtypeStruct(shape=(3,), dtype=f64)
  return jax.pure_callback(func, result_shape_dtype, x)

print(f(x))
# > jaxlib.xla_extension.XlaRuntimeError: INTERNAL: Generated function failed: CpuCallback error: TypeError: func(): incompatible function arguments. The following argument types are supported:
# >    1. func(x: jaxlib.xla_extension.DeviceArray[dtype=float64, shape=(3)]) -> jaxlib.xla_extension.DeviceArray[dtype=float64, shape=(3)]
# >
# > Invoked with types: ndarray
```

<details>
<summary> full output  </summary>

```
Traceback (most recent call last):
  File "/Users/sam/code/jax_nanobind/jax_test.py", line 37, in <module>
    print(f(x))
          ^^^^
  File "/Users/sam/code/jax_nanobind/jax_test.py", line 21, in f
    return jax.pure_callback(func, result_shape_dtype, x)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/jax/_src/callback.py", line 250, in pure_callback_api
    return pure_callback(callback, result_shape_dtypes, *args,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/jax/_src/callback.py", line 192, in pure_callback
    out_flat = pure_callback_p.bind(
               ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/jax/_src/core.py", line 380, in bind
    return self.bind_with_trace(find_top_trace(args), args, params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/jax/_src/core.py", line 383, in bind_with_trace
    out = trace.process_primitive(self, map(trace.full_raise, args), params)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/jax/_src/core.py", line 815, in process_primitive
    return primitive.impl(*tracers, **params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/jax/_src/dispatch.py", line 144, in apply_primitive
    return compiled_fun(*args)
           ^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/jax/_src/profiler.py", line 314, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/jax/_src/interpreters/pxla.py", line 1349, in __call__
    results = self.xla_executable.execute_sharded(input_bufs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
</details>

**Reply by jakevdp** [reply]
Oh, I see, that error seems to be coming from your C++ function. What input types does it accept? `pure_callback` always calls the functions with numpy arrays, not JAX arrays.

**Reply by jakevdp** [reply, marked as answer]
I didn't realize you were trying to call GPU code from JAX. In that case, these docs would be more relevant than `pure_callback`, which is designed for callbacks to the CPU host: https://jax.readthedocs.io/en/latest/Custom_Operation_for_GPUs.html

**Reply by samuelpmish** [reply]
that looks promising, thank you for all the help!
